### PR TITLE
 i#3544 riscv64: Add CI support for RISCV64 

### PR DESCRIPTION
**Pull request by shiptux**
Adds CI support to perform RISCV64 cross compile build. This will
help keep the RISCV64 build healthy by making sure new code
doesn't break anything.

Currently we do not build or run any tests. Added a TODO for that.

Issue #3544

### CODE REVIEW

**Comment by shiptux** [comment]
Could I test the github action workflow in a fork repo?

**Comment by shiptux** [comment]
`https://github.com/DynamoRIO/dynamorio/actions/runs/3741081513/jobs/6350227037`
This CI can currently only compile.

**Comment by derekbruening** [comment]
You can run an action against a branch in the same repo, but only actions already on main show up so a new one can't be tested that way -- but once the file is in place changes to the file can be tested that way.

**Comment by shiptux** [comment]
I have added an github action ci-riscvxx.yml in commit part 2, this same PR.

**Comment by shiptux** [comment]
Those ci-windows workflow are always not pass.Even in other PR.
Another question is the copyright. I saw all of the action files are copyright google. So I keep it consistent. But I don't know if it's right.

**Comment by derekbruening** [comment]
> I had added an github action ci-riscvxx.yml in commit part 2, this same PR.

Dividing into "part 1", "part 2", etc. in the title is for the final merge-to-main commit: we do not use those labels if multiple commits go into the same PR, only for multiple PR's.

**Comment by derekbruening** [comment]
> @[shiptux](https://github.com/shiptux) shiptux [force-pushed](https://github.com/DynamoRIO/dynamorio/compare/134ce476cdb0f29468c527d676838cb4d85d1cd8..fefcc5ee5c46495358c689564e959931a9c36c15) the master branch from [134ce47 ](https://github.com/DynamoRIO/dynamorio/commit/134ce476cdb0f29468c527d676838cb4d85d1cd8)to [fefcc5e ](https://github.com/DynamoRIO/dynamorio/commit/fefcc5ee5c46495358c689564e959931a9c36c15)[yesterday](https://github.com/DynamoRIO/dynamorio/pull/5799#event-8070246369)

Please do not force-push to a shared branch.  It messes up the review history and causes problems if anyone else checked out the branch.  The commit message is cleaned up in the final squash-and-merge so there is no benefit and only negatives to force-pushing a PR.  This is documented in several places: https://dynamorio.org/page_contributing.html, https://dynamorio.org/page_code_reviews.html

**Comment by derekbruening** [comment]
> Those ci-windows workflow are always not pass.Even in other PR.

I think Github updated their Windows VM and caused some problems: someone else hit this in https://github.com/DynamoRIO/dynamorio/pull/5797#issuecomment-1358244930.  One of you should file an issue on this as the first to see it. 

> Another question is the copyright. I saw all of the action files are copyright google. So I keep it consistent. But I don't know if it's right.

Please see https://dynamorio.org/page_contributing.html : "If you would like to contribute code to DynamoRIO, we do not require a formal contributor license agreement or transfer of copyright. Contributions are implicitly assumed to be offered under terms of [DynamoRIO's license](https://dynamorio.org/page_license.html)."

**Comment by derekbruening** [comment]
Please indicate when ready for review.

**Comment by shiptux** [comment]
> > I had added an github action ci-riscvxx.yml in commit part 2, this same PR.
> 
> Dividing into "part 1", "part 2", etc. in the title is for the final merge-to-main commit: we do not use those labels if multiple commits go into the same PR, only for multiple PR's.

I rethought it. That do not need two PR or commit. So i edited the title.

**Comment by shiptux** [comment]
> > @[shiptux](https://github.com/shiptux) shiptux [force-pushed](https://github.com/DynamoRIO/dynamorio/compare/134ce476cdb0f29468c527d676838cb4d85d1cd8..fefcc5ee5c46495358c689564e959931a9c36c15) the master branch from [134ce47 ](https://github.com/DynamoRIO/dynamorio/commit/134ce476cdb0f29468c527d676838cb4d85d1cd8)to [fefcc5e ](https://github.com/DynamoRIO/dynamorio/commit/fefcc5ee5c46495358c689564e959931a9c36c15)[yesterday](https://github.com/DynamoRIO/dynamorio/pull/5799#event-8070246369)
> 
> Please do not force-push to a shared branch. It messes up the review history and causes problems if anyone else checked out the branch. The commit message is cleaned up in the final squash-and-merge so there is no benefit and only negatives to force-pushing a PR. This is documented in several places: https://dynamorio.org/page_contributing.html, https://dynamorio.org/page_code_reviews.html

The last time I did this, to squash two commit. I will not force-push after that.

**Comment by shiptux** [comment]
> Please indicate when ready for review.

i am ready for review

File: suite/runsuite.cmake
Comment (derekbruening) [inline suite/runsuite.cmake:560]: The convention is to always have an issue number for a TODO or XXX comment: `TODO i#3544: `

File: suite/runsuite.cmake
Comment (derekbruening) [inline suite/runsuite.cmake:561]: But the comment just says you can't *run* the tests (hmm actually it says "can't run the build tests": please reword): if you can build them, this option should stay on as we want to test building them.  If you can't build the tests yet, please update the comment.

File: .github/workflows/ci-riscvxx.yml
Comment (derekbruening) [inline .github/workflows/ci-riscvxx.yml:51]: I don't see any tests run under QEMU in this PR: please update the comment.

File: suite/runsuite.cmake
Comment (derekbruening) [inline suite/runsuite.cmake:556]: Please add something like `TODO i#3544: Run tests under QEMU` before this line as the current line is not what we want long-term.

File: suite/runsuite.cmake
Comment (derekbruening) [inline suite/runsuite.cmake:553]: We may want this to behave in the same way as the aarchxx and android cross tests: to be enabled in local suite runs too and not only triggered by this GA CI variable.  The other cross tests will operate locally if the cross compiler is built, so you can run the whole suite on your local machine and include the cross tests.

**Review by derekbruening — COMMENTED** [review]
Thank you for the contribution.  I ended up with a bunch of comments but mostly they are style/nits.  Probably worth another round.  However, I will be traveling and out for a while; maybe @abhinav92003 could do the next round or it's possible I'll be able to at some point in the next week.

File: .github/workflows/ci-riscvxx.yml
Comment (shiptux) [reply .github/workflows/ci-riscvxx.yml:73]: I think that deleting "there are" is better here.

File: .github/workflows/ci-riscvxx.yml
Comment (shiptux) [reply .github/workflows/ci-riscvxx.yml:33]: The "xx" is cover riscv32 and riscv64. I currently have no plans to support riscv32, but maybe someone else will later.

File: .github/workflows/ci-riscvxx.yml
Comment (shiptux) [reply .github/workflows/ci-riscvxx.yml:81]: "sudo add-apt-repository" does a lot of work including "apt update", if I use it, I don't need to do "sudo apt update" again.

File: .github/workflows/ci-riscvxx.yml
Comment (shiptux) [reply .github/workflows/ci-riscvxx.yml:80]: "sudo add-apt-repository" does a lot of work including "apt update", if I use it, I don't need to do "sudo apt update" again.

File: .github/workflows/ci-riscvxx.yml
Comment (shiptux) [reply .github/workflows/ci-riscvxx.yml:51]: Ok. removed that

File: suite/runsuite.cmake
Comment (shiptux) [reply suite/runsuite.cmake:556]: Added it.

File: suite/runsuite.cmake
Comment (shiptux) [reply suite/runsuite.cmake:561]: "#TODO i#3544: Run build tests". Is it better?

File: suite/runsuite.cmake
Comment (shiptux) [reply suite/runsuite.cmake:560]: "#TODO i#3544: Run tests under QEMU"

File: .github/workflows/ci-riscvxx.yml
Comment (shiptux) [reply .github/workflows/ci-riscvxx.yml:79]: Done.

File: .github/workflows/ci-riscvxx.yml
Comment (shiptux) [reply .github/workflows/ci-riscvxx.yml:33]: I removed the "xx". It is hard to see riscv32 hardware.

File: suite/runsuite.cmake
Comment (shiptux) [reply suite/runsuite.cmake:553]: Done.

**Comment by shiptux** [comment]
1 Some modifications:
  2 
  3 Fix typos, indent
  4 Rename riscvxx to riscv64
  5 Use "add-apt-repository" instead.
  6 Fix the "TODO" format.

File: suite/runsuite.cmake
Comment (shiptux) [reply suite/runsuite.cmake:326]: Done

File: suite/runsuite.cmake
Comment (shiptux) [reply suite/runsuite.cmake:554]: Done.

File: suite/runsuite.cmake
Comment (abhinav92003) [inline suite/runsuite.cmake:556]: Style nit: Add space between the `#` and `TODO`: `# TODO i#3544: ...`

File: suite/runsuite.cmake
Comment (abhinav92003) [inline suite/runsuite.cmake:565]: Same here. The space between the `#` and `TODO`.

File: suite/runsuite.cmake
Comment (abhinav92003) [reply suite/runsuite.cmake:566]: It'd be clearer to say

`TODO i#3544: Port tests to RISCV and build them in the workflow`

File: suite/runsuite.cmake
Comment (abhinav92003) [inline suite/runsuite.cmake:425]: This is missing an `AND`.

s/NOT cross_riscv64_linux_only NOT/NOT cross_riscv64_linux_only AND NOT/

**Comment by shiptux** [comment]
Are there developer mail lists?

File: suite/runsuite.cmake
Comment (shiptux) [reply suite/runsuite.cmake]: Done.

File: suite/runsuite.cmake
Comment (shiptux) [reply suite/runsuite.cmake]: Done.

File: suite/runsuite.cmake
Comment (shiptux) [reply suite/runsuite.cmake:566]: > It'd be clearer to say
> 
> `TODO i#3544: Port tests to RISCV and build them in the workflow`
Fine.

**Comment by abhinav92003** [comment]
> Are there developer mail lists?

Yes: https://groups.google.com/g/dynamorio-devs.

File: suite/runsuite.cmake
Comment (shiptux) [reply suite/runsuite.cmake:425]: Done. I will push it later.

File: suite/runsuite.cmake
Comment (abhinav92003) [inline suite/runsuite.cmake:580]: We perhaps also need to restore the optional_cross_compile and run_tests vars (similar to the lines 480 and 482 above).

File: suite/runsuite.cmake
Comment (abhinav92003) [inline suite/runsuite.cmake:558]: This seems like a bug: we're saving prev_run_tests (at line 560 below) *after* modifying it here. Perhaps this should be moved to below L560?

**Review by abhinav92003 — APPROVED** [review]
Sorry for the fragmented review. Found a couple other things we should fix.

Looks good otherwise. Since you don't have committers access, I'll merge the PR after the comments are resolved.

**Comment by abhinav92003** [comment]
The `client.file_io` test is failing on vs2019-32: it's #5802.

**Comment by abhinav92003** [comment]
Two CI workflows are failing. Both due to #5802. Should be safe to merge this PR since it is completely independent.

**Comment by shiptux** [comment]
> > > > Are there developer mail lists?
> > > 
> > > 
> > > Yes: https://groups.google.com/g/dynamorio-devs.
> > 
> > 
> > Could I join this list?
> 
> Of course! Please send a request to join the list on the above link. I don't seem to have admin access to that group, so @derekbruening would need to approve it when he's back.

My email is shiptux@gmail.com.  Thanks.

**Comment by shiptux** [comment]
> > > > > > Are there developer mail lists?
> > > > > 
> > > > > 
> > > > > Yes: https://groups.google.com/g/dynamorio-devs.
> > > > 
> > > > 
> > > > Could I join this list?
> > > 
> > > 
> > > Of course! Please send a request to join the list on the above link. I don't seem to have admin access to that group, so @derekbruening would need to approve it when he's back.
> > 
> > 
> > My email is [shiptux@gmail.com](mailto:shiptux@gmail.com). Thanks.
> 
> I've sent you an invite to join the list.

Sorry for saw it too late, The Github mention send as other mail address. link was expired. Could you please invite me again, Thanks.

**Pull request merged**
